benchmarks/benchmark.py

```diff
@@ -0,0 +1,584 @@
+    :param csv_filename: path to csv file containing results to plot
+    :param reference_csv_filename: path to csv file containing reference results to plot
+    :param out_file: path to image file to store figure into. If it equals to "window" opens it in an interactive window.
+    """
+    import pandas, numpy, matplotlib
+    if out_file != "window":
+        matplotlib.use('Agg')
+    from matplotlib import pyplot as plt
+
+    with open(csv_filename) as f:
+        # google benchmark writes some non-csv data at beginning
+        for line in iter(f.readline, ''):
+            if line.startswith("name,iterations"):
+                f.seek(f.tell() - len(line) - 1, os.SEEK_SET)
+                break
+        data = pandas.read_csv(f)
+    with open(reference_csv_filename) as f:
+        # google benchmark writes some non-csv data at beginning
+        for line in iter(f.readline, ''):
+            if line.startswith("name,iterations"):
+                f.seek(f.tell() - len(line) - 1, os.SEEK_SET)
+                break
+        reference_data = pandas.read_csv(f)
+
+    timing_data = pandas.concat([data["name"].str.split("/", expand=True).iloc[:, :2], data["real_time"]], axis=1)
+    reference_timing_data = pandas.concat(
+        [reference_data["name"].str.split("/", expand=True).iloc[:, :2], reference_data["real_time"]], axis=1)
+    timing_data.columns = reference_timing_data.columns = ["signatures", "sizes", "times"]
+
+    same_in_last_selector = reference_timing_data["signatures"].isin(timing_data["signatures"])
+    reference_timing_data = reference_timing_data.loc[same_in_last_selector, :]
+    assert numpy.all(reference_timing_data["signatures"] == timing_data["signatures"])
+    assert numpy.all(reference_timing_data["sizes"] == timing_data["sizes"])
+
+    timing_data["speedup"] = reference_timing_data["times"] / timing_data["times"]
+    timing_data["sizes"] = timing_data["sizes"].astype(int)
+
+    fig, ax = plt.subplots(figsize=(10, 10))
+    fig.set_tight_layout(True)
+    ax.set_xscale("log")
+    if plot_log_y:
+        ax.set_yscale("log")
+
+    ax.set_xlabel("size")
+    ax.set_ylabel("speedup")
+
+    for n, (signature, sub_data) in enumerate(timing_data.groupby("signatures")):
+        ax.plot(sub_data["sizes"], sub_data["speedup"], "x", color=pick_color(n), label='__nolegend__')
+        avg_sig_speedups = sub_data.groupby(by="sizes")["speedup"].median().reset_index().sort_values(by="sizes")
+        ax.plot(avg_sig_speedups["sizes"], avg_sig_speedups["speedup"], label=signature,
+                color=pick_color(n))
+    plt.plot([1, max(timing_data["sizes"])], [1, 1], "--", color="gray")
+
+    [spine.set_visible(False) for loc, spine in ax.spines.items() if loc in ["top", "right", "left", "bottom"]]
+    ax.minorticks_off()
+    ax.grid()
+    ax.legend()
+    if out_file == "window":
+        plt.show()
+    else:
+        fig.savefig(out_file, bbox_inches="tight")
```

Comment:
I would recommend adding a dpi setting here.

I usually add `dpi=300` but even `dpi=600` depending where I want to use the figure.

The default is 70 (?)